nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -176,11 +179,47 @@ public void checkTypeParameterNullnessForAssignability(Tree tree) {
           typeWithPreservedAnnotations(
               (ParameterizedTypeTree) ((NewClassTree) rhsTree).getIdentifier());
     }
-    if (lhsType != null && rhsType != null) {
+    if (lhsType != null
+        && rhsType != null
+        && lhsType instanceof Type.ClassType
+        && rhsType instanceof Type.ClassType) {
       compareNullabilityAnnotations((Type.ClassType) lhsType, (Type.ClassType) rhsType, tree);
     }
   }
 
+  public void checkTypeParameterNullnessForFunctionReturnType(
+      ExpressionTree retExpr, Symbol.MethodSymbol methodSymbol) {
+    if (!config.isJSpecifyMode()) {
+      return;
+    }
+
+    Type methodType = methodSymbol.getReturnType();
+    // check nullability of parameters only for generics
+    if (methodType.getTypeArguments().length() <= 0) {
+      return;
+    }
+    Type returnExpressionType = ASTHelpers.getType(retExpr);
```

Comment:
From here there is a lot of duplicated code with `checkTypeParameterNullnessForAssignability`.  Let's factor out a common function that takes an `lhsType` (in this case, `methodType`) and an `rhsExpr` (in this case, `retExpr`) and then does the checking, and then call that function from both places.